Add injection to python rest api

ofirZelig opened the pull request:
added a function toRestApi that injects packages

lebauce commented:
run skydive-python-tests

lebauce commented:
ok to test

lebauce commented:
add to whitelist

lebauce commented:
run skydive-python-tests

masco-enovance commented:
it is nice to have some tests and squash the commits.

lebauce commented:
add to whitelist

lebauce commented:
run skydive-python-tests

lebauce commented on contrib/python/api/skydive/rest/client.py:
```diff
@@ -167,3 +167,22 @@ def edgerule_list(self):
+    def injection_create(self,srcTID, dstTID,type="icmp4",payload="",interval=5000,srcIP="", dstIP="",srcMAC="",dstMAC="",count=1,ICMPID=0,srcPort =0, dstPort=0):
+        path = "/api/injectpacket"
+        data = json.dumps({
+        "Src": "G.V().Has('TID', '" + srcTID + "')",
```
nit: please fix indentation

lebauce commented:
@ofirZelig Thank you for this. Could you please also add support for deletion ?

masco-enovance commented:
Please add some tests, it would be easy to validate the code.

safchain commented on contrib/python/api/skydive/rest/client.py:
```diff
@@ -167,3 +167,32 @@ def edgerule_list(self):
+                         count=1, ICMPID=0, srcPort=0, dstPort=0):
+        path = "/api/injectpacket"
+        data = json.dumps({
+            "Src": "G.V().Has('TID', '" + srcTID + "')",
```
I think this has to be adapted

safchain commented on contrib/python/api/skydive/rest/client.py:
```diff
@@ -167,3 +167,32 @@ def edgerule_list(self):
+        path = "/api/injectpacket"
+        data = json.dumps({
+            "Src": "G.V().Has('TID', '" + srcTID + "')",
+            "Dst": "G.V().Has('TID', '" + dstTID + "')",
```
Here as well

ofirZelig commented on contrib/python/api/skydive/rest/client.py:
```diff
@@ -167,3 +167,32 @@ def edgerule_list(self):
         path = "/api/edgerule/%s" % rule_id
         self.request(path, method="DELETE")
+
+    def injection_create(self, srcTID, dstTID, type="icmp4",
```
1 . ok
2. you are right, we will fix it

ofirZelig commented on contrib/python/api/skydive/rest/client.py:
```diff
@@ -167,3 +167,32 @@ def edgerule_list(self):
+    def injection_create(self, srcTID, dstTID, type="icmp4",
+                         payload="", interval=5000, srcIP="",
+                         dstIP="", srcMAC="", dstMAC="",
+                         count=1, ICMPID=0, srcPort=0, dstPort=0):
```
We do not fully understand, can you elaborate?

ofirZelig replied:
Do you mean simply to use the parameter increment?

masco-enovance replied:
yes. It takes a boolean as input

hunchback commented on contrib/python/api/skydive/rest/client.py:
```diff
@@ -167,3 +167,33 @@ def edgerule_list(self):
+                         dstIP="", srcMAC="", dstMAC="",
+                         count=1, ICMPID=0, srcPort=0, dstPort=0):
+
+        path = "/api/injectpacket"
```
this url prefix is shared by all added functions - please define it as a const outside this function blockL

```
INJECT_PATH = "/api/injectpacket"
```

hunchback commented on contrib/python/api/skydive/rest/client.py:
```diff
@@ -167,3 +167,33 @@ def edgerule_list(self):
         path = "/api/edgerule/%s" % rule_id
         self.request(path, method="DELETE")
+
+    def injection_create(self, srcQuery="", dstQuery="", type="icmp4",
```
Python typically uses lowercase_underscore_delimited rather than camelCase for variables so renamed:

for example replaced `srcQuery` with `src_query`, etc.

hunchback commented:
@ofirZelig - please extend the description of what this PR does; and why

hunchback commented:
@ofirZelig - to test your code please use: https://docs.python.org/3/library/unittest.html

the tests should be wrapped by a script which can be invoked via CI:

```
scripts/ci/run-python-api-tests.sh
```

if you need any setup run before running these tests - then implement it in:

```
scripts/ci/install-<XXX>.sh
```

and as a development environment please use a `devmode` VM which can be setup as explained here:

http://skydive.network/documentation/contribute#development-box

ofirZelig commented:
@hunchback 
update: We have cloned the repository to our machine but when building it we recieve the following error message:

[vagrant@dev skydive]$ make
perl: warning: Setting locale failed.
perl: warning: Please check that your locale settings:
    LANGUAGE = (unset),
    LC_ALL = (unset),
    LC_MEASUREMENT = "he_IL.UTF-8",
    LC_PAPER = "he_IL.UTF-8",
    LC_MONETARY = "he_IL.UTF-8",
    LC_NAME = "he_IL.UTF-8",
    LC_ADDRESS = "he_IL.UTF-8",
    LC_NUMERIC = "he_IL.UTF-8",
    LC_TELEPHONE = "he_IL.UTF-8",
    LC_IDENTIFICATION = "he_IL.UTF-8",
    LC_TIME = "he_IL.UTF-8",
    LANG = "en_US.UTF-8"
    are supported and installed on your system.
perl: warning: Falling back to a fallback locale ("en_US.UTF-8").
CC= GOARCH= go get github.com/kardianos/govendor
/home/vagrant/bin/govendor sync
patch -p0 < dpdk/dpdk.govendor.patch
patching file vendor/github.com/intel-go/yanff/low/low.go
rm -rf vendor/github.com/weaveworks/tcptracer-bpf/vendor/github.com/
mv /home/vagrant/.skydive-build-tool bin || true; ln -s vendor src || mv vendor src; cd src/gopkg.in/src-d/proteus.v1/cli/proteus; (unset GOARCH; unset CC; GOPATH=$GOPATH/src/github.com/skydive-project/skydive go install gopkg.in/src-d/proteus.v1/cli/proteus); cd -; mv bin /home/vagrant/.skydive-build-tool; unlink src || mv src vendor; proteus proto -f ${GOPATH}/src -p github.com/skydive-project/skydive/flow/layers
/home/vagrant/src/github.com/skydive-project/skydive
error scanning package "github.com/skydive-project/skydive/flow/layers": /home/vagrant/src/github.com/skydive-project/skydive/flow/layers/layers.go:25:8: could not import github.com/google/gopacket/layers (/home/vagrant/src/github.com/skydive-project/skydive/vendor/github.com/google/gopacket/layers/arp.go:11:2: could not import encoding/binary (/home/vagrant/.gimme/versions/go1.9.1.linux.amd64/src/encoding/binary/binary.go:26:2: could not import io (/home/vagrant/.gimme/versions/go1.9.1.linux.amd64/src/io/pipe.go:12:2: could not import sync (/home/vagrant/.gimme/versions/go1.9.1.linux.amd64/src/sync/cond.go:8:2: could not import sync/atomic (/home/vagrant/.gimme/versions/go1.9.1.linux.amd64/src/sync/atomic/value.go:31:37: cannot convert v (variable of type *Value) to /home/vagrant/.gimme/versions/go1.9.1.linux.amd64/src/unsafe.Pointer)))))
make: *** [Makefile:233: flow/layers/generated.proto] Error 1

safchain commented:
Install by hand is pretty simple, you could try this:
http://skydive.network/documentation/build

lebauce commented:
@ofirZelig It seems your GOPATH is /home/vagrant. Therefore, your GOROOT is inside GOPATH which is known to cause problems. We updated the Vagrant box in this commit 9bb05e888bc9b4fcbb6f9312045fa45c40b55bb1
So you may want to use a newer version of the Vagrant box.

ofirZelig commented:
@safchain  I'm currently working on adding InjectionRule to rules.py, but it is not clear which fileds are optional in the injection. Can you say which is optional from the following fields:
uuid, src, dst, srcip, dstip ,srcmac, dstmac, srcport, dstport, type, payload, trackingid, icmpid, count, interval, increment, starttime

Thank you

ofirZelig commented:
@safchain 
We fixed all the changes which were requested in this PR, and squashed the commits.
The only problem is that in the automatic tests we need to create nodes but they are not able to inject. We are currently trying to solve the problem, do you have an idea what is the best way to get nodes that are able to inject?

safchain commented:
@ofirZelig maybe the loopback interface as source and destination

ofirZelig commented:
@safchain 
The loopback doesnt have a MAC address.
Is there another way?

ofirZelig commented:
@safchain  Is there progress in our PR?

safchain commented:
run skydive-python-tests

ofirZelig commented on contrib/python/api/tests/tests.py:
```diff
@@ -240,7 +240,45 @@ def test_topology_rules(self):
+        quary = "G.V().Has('TID', '" + eth0 + "')"
+        num_injections_before = len(restclient.injection_list())
+        
+        time.sleep(1)
```
why not needed? we need to check if the number of injections is changed

safchain replied:
time.sleep is not needed, please see my comment below

ofirZelig commented on contrib/python/api/tests/tests.py:
```diff
@@ -240,7 +240,45 @@ def test_topology_rules(self):
+        
+        injection_response = restclient.injection_create(quary, quary, count=1000)
+
+        time.sleep(1)
```
what is not needed here?

safchain replied:
It is not needed to wait. I tested without it multiple time without it and it worked.

ofirZelig commented on contrib/python/api/tests/tests.py:
```diff
@@ -240,7 +240,45 @@ def test_topology_rules(self):
         restclient.noderule_delete(noderule1.uuid)
         restclient.noderule_delete(noderule2.uuid)
-        restclient.edgerule_delete(edgerule1.uuid)
+
```
Remove the edge rule deletion or the whitespace?

safchain replied:
I think you updated the PR while I was posting my comment, so forget about this one

safchain replied:
nothing to do my comment is outdated now

safchain commented:
run skydive-python-tests

safchain commented:
@ofirZelig there are multiple coding style issue to fix, please have a look at the CI result

safchain commented:
run skydive-python-tests

safchain commented:
run skydive-python-tests

safchain commented:
run skydive-python-tests

SH118415 commented on contrib/python/api/skydive/packet_injector.py:
```diff
@@ -0,0 +1,103 @@
+        if self.interval:
+            obj["Interval"] = self.interval
+        if self.increment:
+            obj["Increment"] = self.increment
```
return obj

safchain replied:
@SH118415 good catch :)

SH118415 replied:
> @SH118415 good catch :)

@safchain thanks :)

safchain commented:
@ofirZelig why did you reverted this one https://github.com/skydive-project/skydive/pull/1508/commits/3b51fa0a510e5237153fbdc5ce23f08854631f32
? we need to use the allinone mode otherwise it fails

ofirZelig commented:
> In order to be able to test packet injection we need to start an analyzer and an agent however only an analyzer is currently started. In order to have both you need to change
> `analyzer` by `allinone`
> 
> https://github.com/skydive-project/skydive/blob/master/contrib/python/api/tests/tests.py#L51
> 
> Then you could use the `eth0` interface like this `G.V().Has('Name', 'eth0')`
> 
> I tested by hand and it worked so I guess it should work with python

@safchain  
The ssl tests fail when using "allinone" instead of "analyzer", even the tests we did not touch, do you know what is the reason?

ofirZelig commented:
run skydive-python-tests

ofirZelig commented:
> @ofirZelig why did you reverted this one [3b51fa0](https://github.com/skydive-project/skydive/commit/3b51fa0a510e5237153fbdc5ce23f08854631f32)
> ? we need to use the allinone mode otherwise it fails

@safchain 
The ssl tests fail when using "allinone" instead of "analyzer", even the tests we did not touch, do you know what is the reason? Meanwhile we tried a new approach.

safchain commented:
@ofirZelig I'm checking why with TLS it is not working

safchain commented:
@ofirZelig I submitted this #1621 to fix the TLS issue

ofirZelig commented:
@safchain 
I pushed a new version of the tests, hopfully now they would pass.

SH118415 commented on contrib/python/api/tests/tests.py:
```diff
@@ -240,7 +240,35 @@ def test_topology_rules(self):
+        num_injections_after = len(restclient.injection_list())
+
+        self.assertEqual(num_injections_after, num_injections_before + 1,
+                         "injection creation didn's succeed")
```
use didn't instead of didn's

SH118415 commented on contrib/python/api/tests/tests.py:
```diff
@@ -240,7 +240,35 @@ def test_topology_rules(self):
+        num_injections_after_deletion = len(restclient.injection_list())
+
+        self.assertEqual(num_injections_after_deletion, num_injections_before,
+                         "injection deletion didn's succeed")
```
same here

safchain commented:
@ofirZelig can you please rebase on top of master

safchain commented:
run skydive-python-tests

ofirZelig commented:
> @ofirZelig can you please rebase on top of master

Fixed the message, and as I understood, squashing the commits is much easier after the PR is accepted

safchain commented:
@ofirZelig you need to rebase on top of master as we just merged the patch that fixes the TLS/allinone issue otherwise the tests won't pass

ofirZelig commented:
@safchain 
Merged with the current master, everything should be updated

safchain commented:
run skydive-python-tests

ofirZelig commented:
@safchain 
Still the test cant find a node with eth0, even though we merged your commit

safchain commented:
@ofirZelig can you apply this https://paste.fedoraproject.org/paste/naOHo3~ylIlW5B-x9HfTJQ 

Please squash it if it passes I'll merge it. Thanks

lebauce commented:
run skydive-python-tests
